In [241]:
import pandas as pd
import numpy as np

import statsmodels.api as sm
import statsmodels.stats.api as sms
import statsmodels.formula.api as smf

from statsmodels.tsa.seasonal import seasonal_decompose
from statsmodels.graphics.tsaplots import plot_pacf, plot_acf
from statsmodels.tsa.statespace.sarimax import SARIMAX
from statsmodels.tsa.stattools import adfuller

from matplotlib import rc
from matplotlib import rcParams
import matplotlib.pyplot as plt
import seaborn as sns
import scipy.stats as stats

import itertools

In [242]:
data=pd.read_csv('./zillow_data.csv')

In [243]:
data

RegionID  RegionName                 City State              Metro  \
0         84654       60657              Chicago    IL            Chicago   
1         90668       75070             McKinney    TX  Dallas-Fort Worth   
2         91982       77494                 Katy    TX            Houston   
3         84616       60614              Chicago    IL            Chicago   
4         93144       79936              El Paso    TX            El Paso   
...         ...         ...                  ...   ...                ...   
14718     58333        1338             Ashfield    MA    Greenfield Town   
14719     59107        3293            Woodstock    NH          Claremont   
14720     75672       40404                Berea    KY           Richmond   
14721     93733       81225  Mount Crested Butte    CO                NaN   
14722     95851       89155             Mesquite    NV          Las Vegas   

      CountyName  SizeRank   1996-04   1996-05   1996-06  ...  2017-07  \
0           Cook         1  334200.0  335400.0  336500.0  ...  1005500   
1         Collin         2  235700.0  236900.0  236700.0  ...   308000   
2         Harris         3  210400.0  212200.0  212200.0  ...   321000   
3           Cook         4  498100.0  500900.0  503100.0  ...  1289800   
4        El Paso         5   77300.0   77300.0   77300.0  ...   119100   
...          ...       ...       ...       ...       ...  ...      ...   
14718   Franklin     14719   94600.0   94300.0   94000.0  ...   216800   
14719    Grafton     14720   92700.0   92500.0   92400.0  ...   202100   
14720    Madison     14721   57100.0   57300.0   57500.0  ...   121800   
14721   Gunnison     14722  191100.0  192400.0  193700.0  ...   662800   
14722      Clark     14723  176400.0  176300.0  176100.0  ...   333800   

       2017-08  2017-09  2017-10  2017-11  2017-12  2018-01  2018-02  2018-03  \
0      1007500  1007800  1009600  1013300  1018700  1024400  1030700  1033800   
1       310000   312500   314100   315000   316600   318100   319600   321100   
2       320600   320200   320400   320800   321200   321200   323000   326900   
3      1287700  1287400  1291500  1296600  1299000  1302700  1306400  1308500   
4       119400   120000   120300   120300   120300   120300   120500   121000   
...        ...      ...      ...      ...      ...      ...      ...      ...   
14718   217700   218600   218500   218100   216400   213100   209800   209200   
14719   208400   212200   215200   214300   213100   213700   218300   222700   
14720   122800   124600   126700   128800   130600   131700   132500   133000   
14721   671200   682400   695600   695500   694700   706400   705300   681500   
14722   336400   339700   343800   346800   348900   350400   353000   356000   

       2018-04  
0      1030600  
1       321800  
2       329900  
3      1307000  
4       121500  
...        ...  
14718   209300  
14719   225800  
14720   133400  
14721   664400  
14722   357200  

[14723 rows x 272 columns]

In [244]:
#pd.get_option("display.max_rows")
pd.set_option("display.max_rows", 500)

In [245]:
#separate out areas within the state of California
data_ca = data.loc[data.State=='CA']
data_ca.Metro.unique()

array(['San Francisco', 'Los Angeles-Long Beach-Anaheim', 'Sacramento',
       'Napa', 'San Diego', 'Fresno', 'Vallejo', 'Riverside', 'Ventura',
       'San Jose', 'Bakersfield', 'Visalia', 'Hanford', 'Santa Cruz',
       'Modesto', 'Redding', 'Santa Maria-Santa Barbara', 'Stockton',
       'Salinas', 'San Luis Obispo', 'Santa Rosa', 'Chico', 'El Centro',
       'Yuba City', 'Merced', 'Ukiah', 'Truckee', 'Madera', 'Eureka',
       'Susanville', nan, 'Clearlake'], dtype=object)

In [246]:
#separate out the Metro areas of San Francisco 
data_sf = data[(data.Metro == 'San Francisco') & (data.State == 'CA')]
print(data_sf.City.unique())
print(data_sf.shape)
data_sf.head()

['San Francisco' 'Pittsburg' 'Alameda' 'Fremont' 'Hayward' 'Antioch'
 'Brentwood' 'Union City' 'Dublin' 'South San Francisco' 'Martinez'
 'Daly City' 'Oakland' 'Livermore' 'San Leandro' 'Castro Valley'
 'San Mateo' 'Burlingame' 'San Bruno' 'Pleasanton' 'Concord' 'Menlo Park'
 'San Rafael' 'Foster City' 'Richmond' 'Pacifica' 'Pleasant Hill'
 'San Ramon' 'Newark' 'Redwood City' 'Oakley' 'Danville' 'Mill Valley'
 'San Carlos' 'Belmont' 'El Cerrito' 'Lafayette' 'Walnut Creek' 'Berkeley'
 'Novato' 'San Lorenzo' 'El Sobrante' 'Millbrae' 'Hercules' 'Albany'
 'Pinole' 'Orinda' 'Half Moon Bay' 'San Anselmo' 'Tiburon' 'Discovery Bay'
 'Sausalito' 'Alamo' 'Moraga' 'Kentfield' 'Clayton' 'Corte Madera'
 'Fairfax' 'Larkspur' 'Rodeo' 'Portola Valley' 'Atherton' 'Brisbane'
 'Crockett' 'Montara' 'Moss Beach' 'Ross' 'La Honda' 'Woodacre'
 'Stinson Beach']
(134, 272)


RegionID  RegionName           City State          Metro     CountyName  \
9       97564       94109  San Francisco    CA  San Francisco  San Francisco   
27      97771       94565      Pittsburg    CA  San Francisco   Contra Costa   
64      97711       94501        Alameda    CA  San Francisco        Alameda   
76      97565       94110  San Francisco    CA  San Francisco  San Francisco   
186     97743       94536        Fremont    CA  San Francisco        Alameda   

     SizeRank   1996-04   1996-05   1996-06  ...  2017-07  2017-08  2017-09  \
9          10  766000.0  771100.0  776500.0  ...  3767700  3763900  3775000   
27         28  139200.0  138300.0  137500.0  ...   394900   398400   401600   
64         65  222400.0  222300.0  222400.0  ...   965100   975000   987800   
76         77  213900.0  214800.0  216000.0  ...  1503000  1506400  1513500   
186       187  223200.0  223200.0  223300.0  ...   960100   968200   978000   

     2017-10  2017-11  2017-12  2018-01  2018-02  2018-03  2018-04  
9    3799700  3793900  3778700  3770800  3763100  3779800  3813500  
27    405400   408600   410900   413700   417900   424300   430100  
64   1000000  1009900  1021000  1032400  1042000  1053000  1059700  
76   1525200  1533600  1536500  1536200  1544400  1570000  1594000  
186   989600  1002400  1018700  1033100  1045400  1061200  1072700  

[5 rows x 272 columns]

In [247]:
#separate out the Metro areas of San Jose
data_sj = data[(data.Metro == 'San Jose') & (data.State == 'CA')]
print(data_sj.City.unique())
print(data_sj.shape)

['San Jose' 'Santa Clara' 'Milpitas' 'Cupertino' 'Sunnyvale' 'Campbell'
 'Gilroy' 'Morgan Hill' 'Hollister' 'Mountain View' 'Palo Alto' 'Saratoga'
 'Los Gatos' 'Los Altos' 'Stanford' 'Lexington Hills' 'San Martin'
 'San Juan Bautista' 'Aromas']
(55, 272)


# NOTE: 
in order to convert our wide 272 columns dataframe to a long version, we will use the pandas.melt() function:

**pandas.melt(frame, id_vars=None, value_vars=None,
 var_name=None, value_name='value', col_level=None)**
 
value_vars If not specified, uses all columns that are not set as id_vars.

### QUESTION: why the code block below does not show the time and its equal mean values as part of the original dtaframe despite using the melt() function?
if we use 'return melted' instead of 'return melted.groupby('time').aggregate({'value':'mean'})' it gives us the long version of the wide dataframe **but without the averge value**

In [248]:
def melt_data(df):
    melted = pd.melt(df, id_vars=['RegionID','RegionName', 'City', 'State', 'Metro', 
                                  'CountyName','SizeRank'], var_name='time')
    melted['time'] = pd.to_datetime(melted['time'], infer_datetime_format=True, format='%Y-%m-%d')
    melted = melted.dropna(subset=['value'])
    return melted.groupby('time').aggregate({'value':'mean'})

In [249]:
data = melt_data(data)
data

value
time                     
1996-04-01  118299.123063
1996-05-01  118419.044139
1996-06-01  118537.423268
1996-07-01  118653.069278
1996-08-01  118780.254312
1996-09-01  118927.528500
1996-10-01  119120.520316
1996-11-01  119345.352236
1996-12-01  119685.084771
1997-01-01  120121.265712
1997-02-01  120544.241450
1997-03-01  120896.623794
1997-04-01  121268.174510
1997-05-01  121655.897398
1997-06-01  122053.105817
1997-07-01  122456.339057
1997-08-01  122879.926927
1997-09-01  123327.402265
1997-10-01  123823.105590
1997-11-01  124353.737669
1997-12-01  125002.601388
1998-01-01  125736.669833
1998-02-01  126450.778111
1998-03-01  127086.257032
1998-04-01  127736.362972
1998-05-01  128385.014978
1998-06-01  129032.702564
1998-07-01  129676.357127
1998-08-01  130324.519617
1998-09-01  130982.618543
1998-10-01  131678.088697
1998-11-01  132399.342442
1998-12-01  133224.702272
1999-01-01  134133.593921
1999-02-01  135020.881128
1999-03-01  135841.586907
1999-04-01  136683.794842
1999-05-01  137532.147293
1999-06-01  138391.809746
1999-07-01  139265.156718
1999-08-01  140160.604954
1999-09-01  141084.832323
1999-10-01  142070.095711
1999-11-01  143104.040330
1999-12-01  144261.328268
2000-01-01  145522.064733
2000-02-01  146772.265654
2000-03-01  147943.683788
2000-04-01  149107.817637
2000-05-01  150247.760649
2000-06-01  151364.608753
2000-07-01  152457.017608
2000-08-01  153530.934463
2000-09-01  154597.187112
2000-10-01  155691.101045
2000-11-01  156804.237598
2000-12-01  157978.351721
2001-01-01  159201.775407
2001-02-01  160410.849711
2001-03-01  161555.950902
2001-04-01  162664.345730
2001-05-01  163717.366844
2001-06-01  164729.619347
2001-07-01  165707.861474
2001-08-01  166666.391466
2001-09-01  167634.309929
2001-10-01  168653.357200
2001-11-01  169714.371301
2001-12-01  170846.292102
2002-01-01  172058.120845
2002-02-01  173311.938336
2002-03-01  174550.062103
2002-04-01  175769.284723
2002-05-01  176957.558267
2002-06-01  178131.760064
2002-07-01  179301.906919
2002-08-01  180477.891430
2002-09-01  181696.493023
2002-10-01  182962.913714
2002-11-01  184453.159933
2002-12-01  186098.173449
2003-01-01  187610.681669
2003-02-01  188930.160006
2003-03-01  190295.616278
2003-04-01  191691.758603
2003-05-01  193112.990429
2003-06-01  194601.804632
2003-07-01  196006.780152
2003-08-01  197610.872872
2003-09-01  199257.667512
2003-10-01  200939.572619
2003-11-01  202586.613546
2003-12-01  204228.359290
2004-01-01  207492.620700
2004-02-01  209244.217982
2004-03-01  211168.039896
2004-04-01  213296.509107
2004-05-01  215615.416305
2004-06-01  218136.246025
2004-07-01  220658.172765
2004-08-01  223353.453128
2004-09-01  226056.895432
2004-10-01  228744.641697
2004-11-01  231299.465974
2004-12-01  233785.465830
2005-01-01  236197.426127
2005-02-01  238563.124596
2005-03-01  240871.017095
2005-04-01  243402.183594
2005-05-01  245934.664560
2005-06-01  248474.874300
2005-07-01  250549.564286
2005-08-01  252925.442857
2005-09-01  255151.628571
2005-10-01  257204.985714
2005-11-01  258995.542857
2005-12-01  260551.178571
2006-01-01  261477.013375
2006-02-01  262687.485771
2006-03-01  263850.576266
2006-04-01  264959.113546
2006-05-01  265878.713717
2006-06-01  266561.959306
2006-07-01  266695.604630
2006-08-01  266818.043031
2006-09-01  266757.139814
2006-10-01  266590.875524
2006-11-01  266316.523468
2006-12-01  266090.974934
2007-01-01  265816.833298
2007-02-01  265624.760689
2007-03-01  265366.333404
2007-04-01  265036.424874
2007-05-01  264491.781890
2007-06-01  263696.341204
2007-07-01  262677.172218
2007-08-01  261586.201276
2007-09-01  260415.577605
2007-10-01  259139.142452
2007-11-01  257695.776045
2007-12-01  256114.812190
2008-01-01  254395.855767
2008-02-01  252515.797676
2008-03-01  250538.729102
2008-04-01  248491.584018
2008-05-01  246309.457353
2008-06-01  243985.661661
2008-07-01  241710.336283
2008-08-01  239648.205310
2008-09-01  237774.272566
2008-10-01  236009.734513
2008-11-01  234330.385841
2008-12-01  2327

In [250]:
data.value.count()

265